tensorboard/webapp/widgets/line_chart_v2/lib/renderer/svg_renderer.ts

```diff
@@ -0,0 +1,99 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+import {LinePaintOption, ObjectRenderer} from './renderer_types';
+import {Polyline} from '../types';
+import {arePolylinesEqual} from '../utils';
+import {Rect} from '../types';
+
+interface LineCacheValue {
+  type: 'line';
+  data: Polyline;
+  dom: SVGPathElement;
+}
+
+type CacheValue = LineCacheValue;
+
+export class SvgRenderer implements ObjectRenderer<CacheValue> {
+  constructor(private readonly svg: SVGElement) {}
+
+  flush() {
+    // Svg can update the DOM right away when creating the object. No need to flush.
+  }
+
+  onResize(domRect: Rect): void {
+    // Svg viewBox does not need to change with the container size.
+  }
+
+  destroyObject(cachedValue: CacheValue): void {
+    this.svg.removeChild(cachedValue.dom);
+  }
+
+  private createPathDString(polyline: Polyline): string {
+    if (!polyline.length) {
+      return '';
+    }
+
+    const dBuilder: string[] = new Array(polyline.length / 2);
+    dBuilder[0] = `M${polyline[0]},${polyline[1]}`;
+    for (let index = 1; index < polyline.length / 2; index++) {
+      dBuilder[index] = `L${polyline[index * 2]},${polyline[index * 2 + 1]}`;
+    }
+    return dBuilder.join('');
+  }
+
+  createOrUpdateLineObject(
+    cachedLine: CacheValue | null,
+    polyline: Polyline,
+    paintOpt: LinePaintOption
+  ): CacheValue | null {
+    if (!polyline.length) return null;
```

Comment:
If `cachedLine` is an existing cache key with more than zero points,
don’t we still need to do something here to hide the line?

Reply:
Thanks for catching that. It definitely is a bug and I added explicit tests for that.